drivers/input/input_gpio_qdec.c

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_gpio_qdec
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/input/input.h>
+#include <zephyr/kernel.h>
+#include <stdlib.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(input_gpio_qdec, CONFIG_INPUT_LOG_LEVEL);
+
+struct gpio_qdec_config {
+	struct gpio_dt_spec gpio[2];
+	uint32_t sample_time_us;
+	uint32_t idle_timeout_ms;
+	uint16_t axis;
+	uint8_t steps_per_period;
+};
+
+struct gpio_qdec_data {
+	const struct device *dev;
+	struct k_timer sample_timer;
+	uint8_t prev_step;
+	int32_t acc;
+	struct k_work event_work;
+	struct k_work_delayable idle_work;
+	struct gpio_callback gpio_cb;
+};
+
+enum {
+	/* Positive transitions */
+	QDEC_LL_LH = 0x01,
+	QDEC_LH_HH = 0x13,
+	QDEC_HH_HL = 0x32,
+	QDEC_HL_LL = 0x20,
+
+	/* Negative transitions */
+	QDEC_LL_HL = 0x02,
+	QDEC_LH_LL = 0x10,
+	QDEC_HH_LH = 0x31,
+	QDEC_HL_HH = 0x23,
+};
+
+static void gpio_qdec_sample_timer_timeout(struct k_timer *timer)
+{
+	const struct device *dev = k_timer_user_data_get(timer);
+	const struct gpio_qdec_config *cfg = dev->config;
+	struct gpio_qdec_data *data = dev->data;
+	int8_t delta = 0;
+	unsigned int key;
+	uint8_t step = 0x00;
+
+	if (gpio_pin_get_dt(&cfg->gpio[0])) {
+		step |= 0x01;
+	}
+	if (gpio_pin_get_dt(&cfg->gpio[1])) {
+		step |= 0x02;
+	}
+
+	if (data->prev_step == step)
+		return;
+
+	switch ((data->prev_step << 4) | step) {
+	case QDEC_LL_LH:
+	case QDEC_LH_HH:
+	case QDEC_HH_HL:
+	case QDEC_HL_LL:
+		delta = 1;
+		break;
+	case QDEC_LL_HL:
+	case QDEC_LH_LL:
+	case QDEC_HH_LH:
+	case QDEC_HL_HH:
+		delta = -1;
+		break;
+	default:
+		LOG_WRN("%s: lost steps", dev->name);
+	}
+
+	data->prev_step = step;
+
+	key = irq_lock();
+	data->acc += delta;
+	irq_unlock(key);
+
+	if (abs(data->acc) >= cfg->steps_per_period) {
+		k_work_submit(&data->event_work);
+	}
+
+	k_work_reschedule(&data->idle_work, K_MSEC(cfg->idle_timeout_ms));
+}
+
+static void gpio_qdec_event_worker(struct k_work *work)
+{
+	struct gpio_qdec_data *data = CONTAINER_OF(
+			work, struct gpio_qdec_data, event_work);
+	const struct device *dev = data->dev;
+	const struct gpio_qdec_config *cfg = dev->config;
+	unsigned int key;
+	int32_t acc;
+
+	key = irq_lock();
+	acc = data->acc / cfg->steps_per_period;
+	data->acc -= acc * cfg->steps_per_period;
+	irq_unlock(key);
+
+	if (acc != 0) {
+		input_report_rel(data->dev, cfg->axis, acc, true, K_FOREVER);
+	}
+}
+
+static void gpio_qdec_irq_setup(const struct device *dev, bool enable)
+{
+	const struct gpio_qdec_config *cfg = dev->config;
+	unsigned int flags = enable ? GPIO_INT_EDGE_BOTH : GPIO_INT_DISABLE;
+	int ret;
+
+	for (int i = 0; i < 2; i++) {
+		const struct gpio_dt_spec *gpio = &cfg->gpio[i];
```

Comment:
fixed size here (2) sounds like a bad idea

Reply:
Yeah good point, added a `#define GPIO_QDEC_GPIO_NUM 2` and a build check to ensure we have two and only two gpios here.